src/threadpool.c

```diff
@@ -310,3 +338,58 @@ int uv_cancel(uv_req_t* req) {
 
   return uv__work_cancel(loop, req, wreq);
 }
+
+
+static void report(enum estage stage) {
+  QUEUE* q;
+  QUEUE sq;
+  uv_queue_stats_t* s;
+  unsigned length;
+  unsigned threads;
+
+  if (QUEUE_EMPTY(&stats))
+    return;
+
+  length = wq_length;
+  threads = idle_threads;
+
+  QUEUE_MOVE(&stats, &sq);
+
+  while (!QUEUE_EMPTY(&sq)) {
+    q = QUEUE_HEAD(&sq);
+    QUEUE_REMOVE(q);
+    s = QUEUE_DATA(q, struct uv_queue_stats_s, q);
```

Comment:
It would be better to cache `s->submit_cb` and friends in local variables.  (Think: another thread concurrently freeing `s`.)

For the same reason you need to `QUEUE_INSERT_TAIL(&stats, q)` _before_ you release the lock.

Reply:
I see your point about `INSERT_TAIL()`. I'm not seeing the point about `s` being freed, its not allowed to do that until `_stop()` is called, in which case we wouldn't have gotten to this point. Also, `s` is passed as an argument to the callback, so it must be valid memory at point the cb is called, so it seems safe to deref for us, as well as the cb receiving `s` as an argument.

I made both changes: see 0517509

Reply:
The logic w.r.t. to `s` is when another thread calls `uv_queue_stats_stop(s)` and `free(s)` between this thread's calls to `uv_mutex_unlock()` and `s->submit_cb(s, ...)`.